src/interfaces/components/turbine.hpp

```diff
@@ -0,0 +1,299 @@
+
+        // Create shaft base node relative to hub
+        this->shaft_base_node =
+            NodeData(model.AddNode().SetPosition({shaft_length, 0., 0., 1., 0., 0., 0.}).Build());
+
+        // Create yaw bearing node at tower top position
+        this->yaw_bearing_node = NodeData(
+            model.AddNode().SetPosition(tower_top_node.x).SetOrientation({1., 0., 0., 0.}).Build()
+        );
+
+        // Create vector of rotor node IDs (hub, azimuth, shaft base, and blades)
+        std::vector<size_t> rotor_node_ids{
+            this->hub_node.id, this->azimuth_node.id, this->shaft_base_node.id
+        };
+        // Add all blade nodes to rotor_node_ids
+        for (const auto& blade : this->blades) {
+            std::transform(
+                blade.nodes.begin(), blade.nodes.end(), std::back_inserter(rotor_node_ids),
+                [](const NodeData& node) {
+                    return node.id;
+                }
+            );
+        }
+
+        // Loop over blades
+        for (auto i = 0U; i < this->blades.size(); ++i) {
+            // Get element ID of the current blade
+            const auto blade_element_id = this->blades[i].beam_element_id;
+
+            // Rotate each blade to align with global Z-axis
+            model.RotateBeamAboutPoint(blade_element_id, q_x_to_z, origin);
+
+            // Translate blade by hub radius so blade root is at the hub radius
+            model.TranslateBeam(blade_element_id, {0., 0., input.hub_diameter / 2.});
+
+            // Add blade apex node at the origin
+            const auto apex_node_id =
+                model.AddNode().SetPosition({0., 0., 0., 1., 0., 0., 0.}).Build();
+            this->apex_nodes.emplace_back(apex_node_id);
+            rotor_node_ids.push_back(apex_node_id);
+            auto& apex_node = model.GetNode(apex_node_id);
+
+            // Rotate blade for cone angle
+            const auto q_cone = RotationVectorToQuaternion({0., -input.cone_angle, 0.});
+            model.RotateBeamAboutPoint(blade_element_id, q_cone, origin);
+            apex_node.RotateAboutPoint(q_cone, origin);
+
+            // Rotate blade to azimuth angle
+            const auto q_azimuth =
+                RotationVectorToQuaternion({static_cast<double>(i) * blade_angle_delta, 0., 0.});
+            model.RotateBeamAboutPoint(blade_element_id, q_azimuth, origin);
+            apex_node.RotateAboutPoint(q_azimuth, origin);
+        }
+
+        // Create shaft rotation quaternion
+        const auto shaft_rotation = RotationVectorToQuaternion({0., input.shaft_tilt_angle, 0.});
+
+        // Rotate rotor nodes by shaft tilt angle
+        for (const auto& node_id : rotor_node_ids) {
+            model.GetNode(node_id).RotateAboutPoint(shaft_rotation, origin);
+        }
+
+        // Translate rotor to global position
+        for (const auto& node_id : rotor_node_ids) {
+            model.GetNode(node_id).Translate(apex_position);
+        }
+    }
```

Comment:
It is quite a large method and could probably be broken down into more granular methods such as tower nodes, nacelle/shaft nodes, blade nodes, apply transformations etc.